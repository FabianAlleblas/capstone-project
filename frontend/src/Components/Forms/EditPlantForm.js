import PropTypes from 'prop-types'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components/macro'
import { ReactComponent as AddImgIcon } from '../../assets/icons/addimgicon.svg'
import { ReactComponent as ImgDeleteIcon } from '../../assets/icons/imgdeleteicon.svg'
import useForm from '../../hooks/useForm'
import useImageUpload from '../../hooks/useImageUpload'
import Button from '../Buttons/Button'

export default function EditPlantForm({
  updatePlantData,
  deletePlantData,
  plant,
}) {
  EditPlantForm.propTypes = {
    updatePlantData: PropTypes.func.isRequired,
    deletePlantData: PropTypes.func.isRequired,
    plant: PropTypes.object.isRequired,
  }

  const history = useHistory()
  const { handleInputChange, formData } = useForm(plant)
  const { picture, imgData, onChangePicture, deleteImg } = useImageUpload()

  return (
    <Form onSubmit={handleSubmit}>
      <ImgInputWrapper src={imgData}>
        <ImgInput
          name="picture"
          type="file"
          onChange={onChangePicture}
          accept="image/png, image/jpeg"
        />
        {!imgData ? (
          <AddImgIcon />
        ) : (
          <ImgDeleteIconStyled onClick={(event) => deleteImg(event)} />
        )}
      </ImgInputWrapper>
      <Label>
        Your plants name*:
        <Input
          name="name"
          type="text"
          maxLength="28"
          onChange={handleInputChange}
          required
          value={formData?.name}
        />
      </Label>
      <Label>
        The species of your plant*:
        <Input
          name="species"
          type="text"
          maxLength="28"
          onChange={handleInputChange}
          required
          value={formData?.species}
        />
      </Label>
      <Label>
        Special infos:
        <Input
          name="specialInfo"
          type="text"
          maxLength="28"
          onChange={handleInputChange}
          value={formData?.specialInfo ?? ''}
        />
      </Label>
      <ButtonWrapper>
        <ButtonStyled>Update Plant</ButtonStyled>
        <Button onClick={handleCancel} secondaryStyle>
          Cancel
        </Button>
        <Button onClick={handleDelete} secondaryStyle>
          Delete
        </Button>
      </ButtonWrapper>
    </Form>
  )

  function handleSubmit(event) {
    event.preventDefault()
    updatePlantData(formData, plant.id)
    event.target.reset()
    history.push(`/plant?id=${plant.id}`)
  }

  function handleCancel() {
    history.push(`/plant?id=${plant.id}`)
  }

  function handleDelete() {
    deletePlantData(plant.id)
    history.push('/')
  }
}

const Form = styled.form`
  background-color: var(--primary-dark);
  display: grid;
  gap: 20px;
  padding: 0 40px;
`

const ImgInputWrapper = styled.label`
  align-items: center;
  background-image: url(${(props) => props.src});
  background-position: center;
  background-size: cover, contain;
  border-radius: 50%;
  border: ${(props) => !props.src && '4px dashed var(--primary-light)'};
  display: flex;
  height: 148px;
  justify-content: center;
  margin: 0 auto;
  position: relative;
  width: 148px;

  svg {
    margin: 0 4px 4px 0;
  }
`

const ImgDeleteIconStyled = styled(ImgDeleteIcon)`
  position: absolute;
  right: 0;
  top: 0;
`

const ImgInput = styled.input`
  height: 0.1px;
  opacity: 0;
  overflow: hidden;
  position: absolute;
  width: 0.1px;
  z-index: -1;
`

const Label = styled.label`
  color: var(--form-font-color);
  display: flex;
  flex-wrap: wrap;
  font-size: 1.25rem;
`

const Input = styled.input`
  border-radius: 22px;
  border: none;
  color: var(--form-font-color);
  display: block;
  font-family: 'Josefin sans', sans-serif;
  font-size: 1rem;
  margin-top: 8px;
  padding: 14px 20px 10px;
  width: 100%;

  &:focus {
    box-shadow: 0 0 4px var(--shadow) inset;
    outline: none;
  }
`

const ButtonWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 20px;
  padding: 20px 0 0;
`

const ButtonStyled = styled(Button)`
  grid-column: 1/3;
`
