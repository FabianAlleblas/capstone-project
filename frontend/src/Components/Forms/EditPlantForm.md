```js
<EditPlantForm
  updatePlantData={() => {}}
  plant={{ name: 'Bob', species: 'Monstera', specialInfo: 'variegated' }}
/>
```
