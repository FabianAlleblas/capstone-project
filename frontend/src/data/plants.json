[
  {
    "id": 1,
    "name": "Carl",
    "species": "Monstera deliciosa"
  },
  {
    "id": 2,
    "name": "Birgitt",
    "species": "Monstera adansonii"
  },
  {
    "id": 3,
    "name": "Karen",
    "species": "Philodendron gloriosum"
  },
  {
    "id": 4,
    "name": "Jim Jones",
    "species": "Rhaphidophora tetrasperma"
  },
  {
    "id": 5,
    "name": "Samuel",
    "species": "Alocasia zebrina"
  },
  {
    "id": 6,
    "name": "Greg",
    "species": "Scindapsus pictus"
  }
]
