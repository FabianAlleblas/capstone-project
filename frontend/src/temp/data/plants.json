[
  {
    "id": 1,
    "name": "Carl",
    "species": "Monstera deliciosa",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 2,
    "name": "Birgitt",
    "species": "Monstera adansonii",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 3,
    "name": "Karen",
    "species": "Philodendron gloriosum",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 4,
    "name": "Jim Jones",
    "species": "Rhaphidophora tetrasperma",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 5,
    "name": "Samuel",
    "species": "Alocasia zebrina",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 6,
    "name": "Greg",
    "species": "Scindapsus pictus",
    "specialInfo": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  }
]
