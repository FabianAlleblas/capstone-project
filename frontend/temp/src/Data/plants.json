[
  {
    "id": 1,
    "name": "Carl",
    "species": "Monstera deliciosa",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 2,
    "name": "Birgitt",
    "species": "Monstera adansonii",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 3,
    "name": "Karen",
    "species": "Philodendron gloriosum",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 4,
    "name": "Jim Jones",
    "species": "Rhaphidophora tetrasperma",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 5,
    "name": "Samuel",
    "species": "Alocasia zebrina",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  },
  {
    "id": 6,
    "name": "Greg",
    "species": "Scindapsus pictus",
    "info": "variegated",
    "daysLeft": 4,
    "weeksLeft": 3
  }
]
