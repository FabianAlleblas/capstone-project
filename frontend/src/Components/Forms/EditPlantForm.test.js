import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import plants from '../../data/plants.json'
import EditPlantForm from './EditPlantForm'

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

const onSubmitMock = jest.fn()
const onDeleteMock = jest.fn()

const plant = plants[0]

describe('AddPlantForm', () => {
  it('renders with the correct input values', () => {
    const { getByLabelText } = render(
      <EditPlantForm
        updatePlantData={onSubmitMock}
        deletePlantData={onDeleteMock}
        plant={plant}
      />
    )

    expect(getByLabelText('Your plants name*:')).toHaveValue(plant.name)
    expect(getByLabelText('The species of your plant*:')).toHaveValue(
      plant.species
    )
    expect(getByLabelText('Special infos:')).toHaveValue(plant.specialInfo)
  })

  it('calls history.push by clicking the cancel button', () => {
    const { getByText } = render(
      <EditPlantForm
        updatePlantData={onSubmitMock}
        deletePlantData={onDeleteMock}
        plant={plant}
      />
    )

    user.click(getByText('Cancel'))

    expect(mockHistoryPush).toHaveBeenCalled()
  })

  it('calls the right function and history.push by clicking the delete button', () => {
    const { getByText } = render(
      <EditPlantForm
        updatePlantData={onSubmitMock}
        deletePlantData={onDeleteMock}
        plant={plant}
      />
    )

    user.click(getByText('Delete'))

    expect(onDeleteMock).toHaveBeenCalled()
    expect(mockHistoryPush).toHaveBeenCalled()
  })
})
