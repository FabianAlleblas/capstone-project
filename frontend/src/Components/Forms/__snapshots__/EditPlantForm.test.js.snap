// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPlantForm renders correctly 1`] = `
<form
  class="EditPlantForm__Form-bmqh96-0 jbfpuj"
>
  <label
    class="EditPlantForm__ImgInputWrapper-bmqh96-1 ipPaJs"
    data-testid="img-input-wrapper"
    src="http://example.com/example.jpg"
  >
    <input
      accept="image/png, image/jpeg"
      alt="image-input"
      class="EditPlantForm__ImgInput-bmqh96-4 giAAhF"
      name="image"
      type="file"
    />
    <svg>
      addimgicon.svg
    </svg>
  </label>
  <label
    class="EditPlantForm__Label-bmqh96-5 cfWzow"
  >
    Your plants name*:
    <input
      class="EditPlantForm__Input-bmqh96-6 lanFep"
      maxlength="28"
      name="name"
      required=""
      type="text"
      value="Bob"
    />
  </label>
  <label
    class="EditPlantForm__Label-bmqh96-5 cfWzow"
  >
    The species of your plant*:
    <input
      class="EditPlantForm__Input-bmqh96-6 lanFep"
      maxlength="28"
      name="species"
      required=""
      type="text"
      value="Monstera"
    />
  </label>
  <label
    class="EditPlantForm__Label-bmqh96-5 cfWzow"
  >
    Special infos:
    <input
      class="EditPlantForm__Input-bmqh96-6 lanFep"
      maxlength="28"
      name="specialInfo"
      type="text"
      value="variegated"
    />
  </label>
  <div
    class="EditPlantForm__ButtonWrapper-bmqh96-7 gChuOS"
  >
    <button
      class="Button__PrimaryButton-sc-6pjoi6-0 pAbUY EditPlantForm__ButtonStyled-bmqh96-8 bNGiMO"
    >
      Update
    </button>
    <button
      class="Button__SecondaryButton-sc-6pjoi6-1 cSPa-Dk"
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;
